internal/builders/go/pkg/build.go

```diff
@@ -338,43 +355,56 @@ func isAllowedEnvVariable(name string) bool {
 // TODO: maybe not needed if handled directly by go compiler.
 func (b *GoBuild) generateLdflags() (string, error) {
 	var a []string
+	reVar := regexp.MustCompile(`{{ \.([A-Z][a-z]*) }}`)
+	reDyn := regexp.MustCompile(`{{ \.Env\.(\w+) }}`)
 
+	// Resolve variables.
 	for _, v := range b.cfg.Ldflags {
-		var res string
-		ss := "{{ .Env."
-		es := "}}"
-		found := false
-		for {
-			start := strings.Index(v, ss)
-			if start == -1 {
-				break
-			}
-			end := strings.Index(string(v[start+len(ss):]), es)
-			if end == -1 {
-				return "", fmt.Errorf("%w: %s", errorInvalidEnvArgument, v)
-			}
 
-			name := strings.Trim(string(v[start+len(ss):start+len(ss)+end]), " ")
-			if name == "" {
-				return "", fmt.Errorf("%w: %s", errorEnvVariableNameEmpty, v)
+		// Special variables.
+		names := reVar.FindAllString(v, -1)
```

Comment:
You could use FindAllStringSubmatch and not have to do the ReplaceAll calls below.

Reply:
I don't think I can because the match is on `KEYWORD` in `{{ .KEYWORD }}` but we want to replace the entire string. Let me know if I'm missing something though

Reply:
FindAllStringMatch gives you an array of full match and submatches. In the case of reVar you'll get something like `[["{{ .Env }}", "Env"] ["{{ .Version }}", "Version"], ...]`
https://go.dev/play/p/LnezjiVtBp3

Reply:
Thanks. SO if the match occurs multiple times, eg `[[{{ .Version }} Version] [{{ .Version }} Version]]`, I still need to iterate over them and replace them one by one, I think.
And I also need to parse each match to know if it's a full match or not (well, their position is always odd so that helps)

I think I'll keep as is for now, if you don't mind.

I would have liked to use the `re. ReplaceAllString`

Reply:
Ok, The code works.

But just be clear I would have written it like this. You get the full value of with brackets in `n[0]` and just the name in `n[1]`.
```
names := reVar.FindAllStringSubmatch(v, -1)
for _, n := range names {
	// {{ .Version }} variable.
	if n[1] != "Version" {
		return "", fmt.Errorf("%w: %s", errorInvalidEnvArgument, n[1])
	}

	version := getVersion()
	v = strings.ReplaceAll(v, n[0], version)
}
```

It still works if there are multiple matches with the same name though a bit inefficient because ReplaceAll will have done the work on the first time through.
https://go.dev/play/p/U0jiQ2rMab1